EIPS/eip-6046.md

```diff
@@ -0,0 +1,63 @@
+---
+eip: 6046
+title: Replace SELFDESTRUCT with DEACTIVATE
+description: Change SELFDESTRUCT to not delete storage keys and use a special value in the account nonce to signal deactivation
+author: Alex Beregszaszi (@axic)
+discussions-to: https://ethereum-magicians.org/t/almost-self-destructing-selfdestruct-deactivate/11886
+status: Draft
+type: Standards Track
+category: Core
+created: 2022-11-25
+requires: 2681
+---
+
+## Abstract
+
+Change `SELFDESTRUCT` to not delete all storage keys, and to use a special value in the account nonce to signal *decativated* accounts. Because the semantics of revival change (storage keys may exists), we also rename the instruction to `DEACTIVATE`.
+
+## Motivation
+
+The `SELFDESTRUCT` instruction currently has a fixed price, but is unbounded in terms of how many storage/account changes it performs (it needs to delete all keys). This is an outstanding concern for long.
+
+Furthermore with *Verkle trees* accounts will be organised differently: account properties, including storage, will have individual keys, and it will not be possible to traverse and find all used keys. This makes `SELFDESTRUCT` very challenging to support in Verkle trees.
+
+## Specification
+
+1. Change the rules introduced by [EIP-2681](./eip-2681.md) such that regular nonce increase is bound by `2^64-2` instead of `2^64-1`. This applies from genesis.
+
+2. The behaviour of `SELFDESTRUCT` is changed:
+
+  - Do not delete any storage keys and also leave the account in place.
+  - Do not give a refund.
+  - Transfer all value and set account balance to `0.`
+  - Set the account nonce to `2^64-1`.
+
+2. Modify account execution (triggered both via external transactions or CALL*), such that execution succeeds and returns an empty buffer if the nonce equals `2^64-1`.
+
+  - Note that the account can still receive non-executable value transfers (such as coinbase transactions or other `SELFDESTRUCT`s).
+
+3. Modify `CREATE2` such that it allows account creation if the nonce equals `2^64-1`.
+
+  - Note that a successful `CREATE2` will change the account code, nonce and potentially balance.
+
+4. Rename the `SELFDESTRUCT` instruction to `DEACTIVATE`, because the semantics of "account revival" are changed: the old storage items will remain, and newly deployed code must be aware of this.
+
+## Rationale
+
+There have been various proposals of removing `SELFDESTRUCT` and many would just outright remove the deletion capability. This breaks certain usage patterns, which the *deactivation* option leaves intact, albeit with minor changes. This only affects *newly* deployed code, and not existing one.
```

Comment:
> There have been various proposals of removing `SELFDESTRUCT`

Might be worth listing a few.

Reply:
Supposedly we cannot link to external resources, so why bother making up a weird list without links?

Reply:
You can refer to other EIPs.

Reply:
A number of the previous attempts were not merged as EIPs, only exists as PRs. They can't be linked.

Reply:
*Basically pointing out how bad our EIP-1 system is.*

Reply:
Final EIPs can link to non-final EIPs. They just can't do so in the normative sections. The rationale section is non-normative.